### PR TITLE
[feat] - Make skipping binaries configurable

### PR DESCRIPTION
**Pull request by ahrav**
### Description:
Add option to disable scanning of binary files from Git sources

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/handlers/archive.go [hunk 1]
```diff
@@ -115,6 +119,33 @@ func (a *Archive) openArchive(ctx logContext.Context, depth int, reader io.Reade
...
+		}
+
+		// Create a new reader that starts with the buffer we just read
+		// and continues with the rest of the original reader.
+		reader = io.MultiReader(bytes.NewReader(buffer[:n]), reader)
+
+		mime := mimetype.Detect(buffer)
+		mimeT := mimeType(mime.String())
+
+		fmt.Println(mimeT)
```
Comment (mcastorina) [inline pkg/handlers/archive.go:136]: Debug print?

File: pkg/handlers/archive.go [hunk 2]
```diff
@@ -115,6 +119,33 @@ func (a *Archive) openArchive(ctx logContext.Context, depth int, reader io.Reade
...
+
+		if mimeT == machOType || mimeT == octetStream {
+			ctx.Logger().V(5).Info("skipping binary file", "ext", mimeT)
+			return nil
+		}
+
+		if common.SkipFile(mime.Extension()) {
+			ctx.Logger().V(5).Info("skipping file", "ext", mimeT)
+			return nil
+		}
```
Comment (mcastorina) [inline pkg/handlers/archive.go:146]: Why do we do this only if `skipBinaries` is set?

File: pkg/handlers/archive.go [hunk 3]
```diff
@@ -161,9 +192,11 @@ func (a *Archive) extractorHandler(archiveChan chan []byte) func(context.Context
...
 
-		if common.SkipFile(f.Name()) {
-			lCtx.Logger().V(5).Info("skipping file", "filename", f.Name())
-			return nil
+		if a.skipBinaries {
+			if common.SkipFile(f.Name()) {
+				lCtx.Logger().V(5).Info("skipping file", "filename", f.Name())
+				return nil
+			}
 		}
```
Comment (mcastorina) [inline pkg/handlers/archive.go:200]: Same question here, why was this changed to only skip files if `skipBinaries` is set?

File: pkg/handlers/archive.go [hunk 4]
```diff
@@ -222,6 +255,8 @@ type mimeType string
 const (
 	arMimeType  mimeType = "application/x-unix-archive"
 	rpmMimeType mimeType = "application/x-rpm"
+	machOType   mimeType = "application/x-mach-binary"
+	octetStream mimeType = "application/octet-stream"
```
Comment (mcastorina) [inline pkg/handlers/archive.go:259]: What about ELF or EXE files? Are those covered by `application/octet-stream`?

File: pkg/sources/git/git.go [hunk 5]
```diff
@@ -58,14 +58,15 @@ type Git struct {
...
+	skipBinaries       bool
 }
 
 type metrics struct {
 	commitsScanned uint64
 }
 
 func NewGit(sourceType sourcespb.SourceType, jobID sources.JobID, sourceID sources.SourceID, sourceName string, verify bool, concurrency int,
-	sourceMetadataFunc func(file, email, commit, timestamp, repository string, line int64) *source_metadatapb.MetaData,
+	sourceMetadataFunc func(file, email, commit, timestamp, repository string, line int64) *source_metadatapb.MetaData, skipBinaries bool,
```
Comment (mcastorina) [inline pkg/sources/git/git.go:69]: This function just keeps growing 🤣 

Not something to fix in this PR of course, but I'm looking forward to refactoring this.

File: pkg/handlers/archive.go (hunk 2 above)
Comment (ahrav) [reply pkg/handlers/archive.go:146]: Oh i forgot, we had added a bunch of binary extensions to that function, but that probably doesn't make much sense anymore. I'll update that function to remove binary extensions.

File: pkg/handlers/archive.go (hunk 3 above)
Comment (ahrav) [reply pkg/handlers/archive.go]: updated

File: pkg/handlers/archive.go (hunk 4 above)
Comment (ahrav) [reply pkg/handlers/archive.go:260]: Some of the reading i did they showed up as octet-stream, but i also read that they can be `application-x-elf`. Just to be safe, I added additional extensions to the list of known binary extensions. There might be a more robust way of determining if the file is a binary format.

File: pkg/sources/git/git.go (hunk 5 above)
Comment (ahrav) [reply pkg/sources/git/git.go:69]: Yea... it was a bit painful adding another param :(

File: pkg/common/vars.go [hunk 6]
```diff
@@ -58,3 +90,13 @@ func SkipFile(filename string) bool {
...
+
+// IsBinary returns true if the file extension is in the binaryExtensions list.
+func IsBinary(filename string) bool {
+	for _, ext := range binaryExtensions {
+		if strings.TrimPrefix(filepath.Ext(filename), ".") == ext {
+			return true
+		}
+	}
+	return false
+}
```
Comment (mcastorina) [inline pkg/common/vars.go:102]: **Nit:** In another PR we should do a map lookup (or even better?, a switch case).

File: pkg/common/vars.go [hunk 7]
```diff
@@ -58,3 +90,13 @@ func SkipFile(filename string) bool {
 	}
 	return false
 }
+
+// IsBinary returns true if the file extension is in the binaryExtensions list.
+func IsBinary(filename string) bool {
+	for _, ext := range binaryExtensions {
+		if strings.TrimPrefix(filepath.Ext(filename), ".") == ext {
```
Comment (mcastorina) [inline pkg/common/vars.go:97]: **Nit:** case-insensitive comparison

```go
fileExt := strings.TrimPrefix(filepath.Ext(filename), ".")
if strings.EqualFold(ext, fileExt) {
    // ...
}
```

File: pkg/handlers/archive.go [hunk 8]
```diff
@@ -115,6 +119,31 @@ func (a *Archive) openArchive(ctx logContext.Context, depth int, reader io.Reade
...
 func (a *Archive) handleNonArchiveContent(ctx logContext.Context, reader io.Reader, archiveChan chan []byte) error {
+	buffer := make([]byte, 512)
+	n, err := reader.Read(buffer)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return fmt.Errorf("unable to read file for MIME type detection: %w", err)
+	}
+
+	// Create a new reader that starts with the buffer we just read
+	// and continues with the rest of the original reader.
+	reader = io.MultiReader(bytes.NewReader(buffer[:n]), reader)
```
Comment (mcastorina) [inline pkg/handlers/archive.go:130]: Would using [bufio.Peek](https://pkg.go.dev/bufio#Reader.Peek) be equivalent here? I find that a little easier to understand than reading a buffer then resetting it with a multireader.

```go
bufReader := bufio.NewReaderSize(reader, 512)
buffer, err := bufReader.Peek(512)
if err != nil && !errors.Is(err, io.EOF) {
    // ...
}
```

File: pkg/handlers/archive.go [hunk 9]
```diff
@@ -115,6 +119,31 @@ func (a *Archive) openArchive(ctx logContext.Context, depth int, reader io.Reade
...
+
+	mime := mimetype.Detect(buffer)
+	mimeT := mimeType(mime.String())
+
+	if common.SkipFile(mime.Extension()) {
+		ctx.Logger().V(5).Info("skipping file", "ext", mimeT)
+		return nil
+	}
```
Comment (mcastorina) [inline pkg/handlers/archive.go:138]: Does this actually work? `SkipFile` expects a full filename and we're just passing the extension.

**Review by mcastorina — APPROVED** [review]

File: pkg/handlers/archive.go (hunk 9 above)
Comment (ahrav) [reply pkg/handlers/archive.go:138]: According to the documentation `.Extension()` returns:
```
// Extension returns the file extension associated with the MIME type.
// It includes the leading dot, as in ".html". When the file format does not
// have an extension, the empty string is returned.
```

File: pkg/common/vars.go (hunk 6 above)
Comment (ahrav) [reply pkg/common/vars.go:102]: nah, i'm fix this right now lol. I like the map too. I'll also post benchies.

Comment (ahrav) [reply pkg/common/vars.go:102]: Screenshot 2023-12-15 at 10 41 03 AM

#microptimizations 🥳

File: pkg/handlers/archive.go (hunk 8 above)
Comment (ahrav) [reply pkg/handlers/archive.go:130]: beautiful, thank you!

Comment (ahrav) [reply pkg/handlers/archive.go]: Ahh weird... with that change the tests fail. I'll dig into why.

Comment (mcastorina) [reply pkg/handlers/archive.go]: Hm.. did you replace all the `reader` variables with `bufReader`? That was the only downside I could think of with this approach.

Comment (ahrav) [reply pkg/handlers/archive.go]:
```
// Peek returns the next n bytes without advancing the reader. The bytes stop
// being valid at the next read call
```

Comment (mcastorina) [reply pkg/handlers/archive.go]: I take that to mean the slice returned by `Peek` becomes invalid (since it's probably returning the internal buffer to avoid a copy). Do we need the buffer to remain valid after the next read?

File: proto/sources.proto [hunk 10]
```diff
@@ -226,6 +229,7 @@ message GitHub {
   bool includePullRequestComments = 14;
   bool includeIssueComments = 15;
   bool includeGistComments = 16;
+  bool skip_binaries = 17;
```
Comment (rgmz) [inline proto/sources.proto:232]: Slightly off topic but what is the convention for protos? Snake case, camel case, both?

Comment (ahrav) [reply proto/sources.proto:232]: Argh, i should've probably just used camel case, even though the style guide does recommend snake.
https://protobuf.dev/programming-guides/style/#message-field-names

Comment (ahrav) [reply proto/sources.proto:232]: I think I can actually update all of them to be snake_case. I'm fairly certain that wouldn't be a breaking change. I'll double check though.

Comment (mcastorina) [reply proto/sources.proto:232]: I thought the idea with protos was that the name doesn't matter, just the number. So I agree changing them shouldn't be a breaking change. The generated Go code would probably be the same too (hopefully).

Comment (rgmz) [reply proto/sources.proto:232]: Fair enough. I was curious because I was about to add a new field and realized there was an inconsistency.

Comment (mcastorina) [reply proto/sources.proto:232]: Oh, yeah. I would also like to know (and stick to) whatever the convention is! Just commenting on the renaming idea.

Thanks for pointing out the inconsistency!
